tests/functional/olp-cpp-sdk-dataservice-read/DataserviceReadVolatileLayerClientTest.cpp

```diff
@@ -224,4 +341,273 @@ TEST_F(VolatileLayerClientTest, GetPartitionsDifferentFetchOptions) {
+                         .WithTileKeys(tile_keys)
+                         .WithMinLevel(0)
+                         .WithMaxLevel(-1);
+
+      std::promise<PrefetchTilesResponse> promise;
+      std::future<PrefetchTilesResponse> future = promise.get_future();
+      auto token = client.PrefetchTiles(
+          request, [&promise](PrefetchTilesResponse response) {
+            promise.set_value(std::move(response));
+          });
+
+      ASSERT_NE(future.wait_for(kTimeout), std::future_status::timeout);
+      PrefetchTilesResponse response = future.get();
+      EXPECT_TRUE(response.IsSuccessful());
+      const auto& result = response.GetResult();
+
+      for (auto tile_result : result) {
+        EXPECT_TRUE(tile_result->IsSuccessful());
+        ASSERT_TRUE(tile_result->tile_key_.IsValid());
+      }
+
+      ASSERT_EQ(tile_keys.size(), result.size());
+    }
+
+    {
+      SCOPED_TRACE(" min level greater than max level");
+      auto request = olp::dataservice::read::PrefetchTilesRequest()
+                         .WithTileKeys(tile_keys)
+                         .WithMinLevel(12)
+                         .WithMaxLevel(10);
+
+      std::promise<PrefetchTilesResponse> promise;
+      std::future<PrefetchTilesResponse> future = promise.get_future();
+      auto token = client.PrefetchTiles(
+          request, [&promise](PrefetchTilesResponse response) {
+            promise.set_value(std::move(response));
+          });
+
+      ASSERT_NE(future.wait_for(kTimeout), std::future_status::timeout);
+      PrefetchTilesResponse response = future.get();
+      EXPECT_TRUE(response.IsSuccessful());
+      const auto& result = response.GetResult();
+
+      for (auto tile_result : result) {
+        EXPECT_TRUE(tile_result->IsSuccessful());
+        ASSERT_TRUE(tile_result->tile_key_.IsValid());
+      }
+
+      ASSERT_EQ(tile_keys.size(), result.size());
+    }
+
+    {
+      SCOPED_TRACE(" min level is zero");
+      auto request = olp::dataservice::read::PrefetchTilesRequest()
+                         .WithTileKeys(tile_keys)
+                         .WithMinLevel(0)
+                         .WithMaxLevel(3);
+
+      std::promise<PrefetchTilesResponse> promise;
+      std::future<PrefetchTilesResponse> future = promise.get_future();
+      auto token = client.PrefetchTiles(
+          request, [&promise](PrefetchTilesResponse response) {
+            promise.set_value(std::move(response));
+          });
+
+      ASSERT_NE(future.wait_for(kTimeout), std::future_status::timeout);
+      PrefetchTilesResponse response = future.get();
+      EXPECT_TRUE(response.IsSuccessful());
+      const auto& result = response.GetResult();
+
+      for (auto tile_result : result) {
+        EXPECT_TRUE(tile_result->IsSuccessful());
+        ASSERT_TRUE(tile_result->tile_key_.IsValid());
+      }
+
+      ASSERT_EQ(tile_keys.size(), result.size());
+    }
+  }
+}
+
+TEST_F(VolatileLayerClientTest, PrefetchWithCancellableFuture) {
```

Comment:
You are testing too much. You don't need to replicate the integration tests here. You only need to test the real functionality with OLP, resp. with the mocked server later on. So that you have covered the e2e scenario.
You don't need to replicate the entire integration tests here. This is not the point.

Reply:
Removed.
Actually, I took them from VersionedLayerClient prefetch fv tests. So they need cleanup too.